waspc/tests/Generator/Valid/PackageJsonTest.hs

```diff
@@ -0,0 +1,183 @@
+    makeRequiredDepValidator,
+  )
+import qualified Wasp.Generator.Valid.Validator as V
+
+spec_PackageJson :: Spec
+spec_PackageJson = do
+  describe "makeOptionalDepValidator" $ do
+    it "succeeds when optional dependency is not present" $
+      forEachCase $ \depType pkgJson -> do
+        testSuccess
+          (makeOptionalDepValidator depType ("optional-pkg", "1.0.0"))
+          pkgJson
+
+    it "succeeds when optional dependency has correct version" $
+      forEachCase $ \depType pkgJson -> do
+        testSuccess
+          (makeOptionalDepValidator depType ("optional-pkg", "1.0.0"))
+          (pkgJson `withDep` (depType, ("optional-pkg", "1.0.0")))
+
+    it "fails when optional dependency has wrong version" $
+      forEachCase $ \depType pkgJson -> do
+        testFailure
+          (makeOptionalDepValidator depType ("optional-pkg", "1.0.0"))
+          (pkgJson `withDep` (depType, ("optional-pkg", "2.0.0")))
+          ["Wasp requires package \"optional-pkg\" to be version \"1.0.0\" if present."]
+
+    it "sets correct field path for dependencies" $
+      forEachCase $ \depType pkgJson -> do
+        let fieldName = depTypeToFieldName depType
+
+        testFieldPath
+          (makeOptionalDepValidator depType ("optional-pkg", "1.0.0"))
+          (pkgJson `withDep` (depType, ("optional-pkg", "2.0.0")))
+          [fieldName, "optional-pkg"]
+
+  describe "makeRequiredDepValidator" $ do
+    it "succeeds when required dependency is present with correct version" $
+      forEachCase $ \depType pkgJson -> do
+        testSuccess
+          (makeRequiredDepValidator depType ("required-pkg", "1.0.0"))
+          (pkgJson `withDep` (depType, ("required-pkg", "1.0.0")))
+
+    it "fails when required dependency is missing" $
+      forEachCase $ \depType pkgJson -> do
+        testFailure
+          (makeRequiredDepValidator depType ("required-pkg", "1.0.0"))
+          pkgJson
+          ["Wasp requires package \"required-pkg\" with version \"1.0.0\"."]
+
+    it "fails when required dependency has wrong version" $
+      forEachCase $ \depType pkgJson -> do
+        testFailure
+          (makeRequiredDepValidator depType ("required-pkg", "1.0.0"))
+          (pkgJson `withDep` (depType, ("required-pkg", "2.0.0")))
+          ["Wasp requires package \"required-pkg\" to be version \"1.0.0\"."]
+
+    it "fails when required dependency is in the opposite list" $
+      forEachCase $ \depType pkgJson -> do
+        let oppositeDepType = depTypeToOpposite depType
+        let fieldName = depTypeToFieldName depType
+
+        testFailure
+          (makeRequiredDepValidator depType ("required-pkg", "1.0.0"))
+          (pkgJson `withDep` (oppositeDepType, ("required-pkg", "1.0.0")))
+          ["Wasp requires package \"required-pkg\" to be in \"" <> fieldName <> "\"."]
+
+    it "sets correct field path" $
+      forEachCase $ \depType pkgJson -> do
+        let fieldName = depTypeToFieldName depType
+
```

Comment:
Subjective, but when code is so short, having a newline like this makes it harder ot read, because now it looks split into two halves, and it takes me a moment to figure out it is still one block. For bigger blocks sure, but for short stuff like this one, I would rather it is compact. Goes for the rest of the file also. Do as you want though, RAW.